File: leetcode/235.二叉搜索树的最近公共祖先.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    x := root.Val
    if p.Val < x && q.Val < x {
        return lowestCommonAncestor(root.Left, p, q)
    }
    if p.Val > x && q.Val > x {
        return lowestCommonAncestor(root.Right, p, q)
    }
    return root
}

File: leetcode/235.二叉搜索树的最近公共祖先.py
class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        x = root.val
        if p.val < x and q.val < x:
            return self.lowestCommonAncestor(root.left, p, q)
        if p.val > x and q.val > x:
            return self.lowestCommonAncestor(root.right, p, q)
        return root